types/hapi__joi/index.d.ts

```diff
@@ -19,1378 +19,1883 @@
+         * the array item being tested, not the value of the `ref` target.
+         */
+        has(schema: SchemaLike): this;
+
+        /**
+         * List the types allowed for the array values.
+         * If a given type is .required() then there must be a matching item in the array.
+         * If a type is .forbidden() then it cannot appear in the array.
+         * Required items can be added multiple times to signify that multiple items must be found.
+         * Errors will contain the number of items that didn't match.
+         * Any unmatched item having a label will be mentioned explicitly.
+         *
+         * @param type - a joi schema object to validate each array item against.
+         */
+        items(...types: SchemaLike[]): this;
+
+        /**
+         * Specifies the exact number of items in the array.
+         */
+        length(limit: number | Reference): this;
+
+        /**
+         * Specifies the maximum number of items in the array.
+         */
+        max(limit: number | Reference): this;
+
+        /**
+         * Specifies the minimum number of items in the array.
+         */
+        min(limit: number | Reference): this;
+
+        /**
+         * Lists the types in sequence order for the array values where:
+         * @param type - a joi schema object to validate against each array item in sequence order. type can be multiple values passed as individual arguments.
+         * If a given type is .required() then there must be a matching item with the same index position in the array.
+         * Errors will contain the number of items that didn't match.
+         * Any unmatched item having a label will be mentioned explicitly.
+         */
+        ordered(...types: SchemaLike[]): this;
+
+        /**
+         * Allow single values to be checked against rules as if it were provided as an array.
+         * enabled can be used with a falsy value to go back to the default behavior.
+         */
+        single(enabled?: any): this;
+
+        /**
+         * Allow this array to be sparse.
+         * enabled can be used with a falsy value to go back to the default behavior.
+         */
+        sparse(enabled?: any): this;
+
+        /**
+         * Sorts the array by given order.
+         */
+        sort(options?: ArraySortOptions): this;
+
+        /**
+         * Requires the array values to be unique.
+         * Remember that if you provide a custom comparator function,
+         * different types can be passed as parameter depending on the rules you set on items.
+         * Be aware that a deep equality is performed on elements of the array having a type of object,
+         * a performance penalty is to be expected for this kind of operation.
+         */
+        unique(comparator?: string, options?: ArrayUniqueOptions): this;
+        unique<T = any>(comparator?: (a: T, b: T) => boolean, options?: ArrayUniqueOptions): this;
```

Comment:
While we are here, we might want to consider making both of these `unknown` instead.

Reply:
The only downside to it is we will need to specify the type when passing the comparator function or else: 
![image](https://user-images.githubusercontent.com/7302877/65016944-cd20b180-d925-11e9-9bd2-9ceafbbc6d86.png)

What do you think? Enforcing to add the type to the comparator signature seems fine by me

Reply:
Seems like upgrading to TS 3.0 would require other packages to upgrade as well:

https://travis-ci.org/DefinitelyTyped/DefinitelyTyped/builds/585918998#L237
```
Error: hapi__hapi depends on hapi__joi but has a lower required TypeScript version.
```